apps/ejabberd/src/ejabberd_s2s_in.erl

```diff
@@ -832,3 +720,132 @@ match_labels([DL | DLabels], [PL | PLabels]) ->
         false ->
             false
     end.
+
+get_sasl(SockMod, Socket, TLSCertverify) ->
+    case SockMod:get_peer_certificate(Socket) of
+        {ok, Cert} ->
+            case SockMod:get_verify_result(Socket) of
+                0 ->
+                    [#xmlel{name = <<"mechanisms">>,
+                            attrs = [{<<"xmlns">>, ?NS_SASL}],
+                            children = [#xmlel{name = <<"mechanism">>,
+                                               children = [#xmlcdata{content = <<"EXTERNAL">>}]}]}];
+                CertVerifyRes ->
+                    check_sasl_tls_certveify(TLSCertverify, CertVerifyRes, Cert)
+            end;
+        error ->
+            []
+    end.
+
+check_sasl_tls_certveify(true, CertVerifyRes, Cert) ->
+    {error_cert_verif, CertVerifyRes, Cert};
+check_sasl_tls_certveify(false, _, _) ->
+    [].
+
+get_auth_res({ok, Cert}, AuthDomain, StateData) ->
+    case (StateData#state.sockmod):get_verify_result(
+           StateData#state.socket) of
+        0 ->
+            check_auth_domain(AuthDomain, Cert);
+        _ ->
+            false
+    end;
+get_auth_res(_, _, _) ->
+    false.
+
+check_auth_domain(error, _) ->
+    false;
+check_auth_domain(AuthDomain, Cert) ->
+    case ejabberd_s2s:domain_utf8_to_ascii(AuthDomain) of
+        false ->
+            false;
+        PCAuthDomain ->
+            lists:any(
+              fun(D) ->
+                      match_domain(
+                        PCAuthDomain, D)
+              end, get_cert_domains(Cert))
+    end.
+
+handle_auth_res(true, AuthDomain, StateData) ->
+    send_element(StateData,
+                 #xmlel{name = <<"success">>,
+                        attrs = [{<<"xmlns">>, ?NS_SASL}]}),
+    ?DEBUG("(~w) Accepted s2s authentication for ~s",
+           [StateData#state.socket, AuthDomain]),
+    {next_state, wait_for_stream,
+     StateData#state{streamid = new_id(),
+                     authenticated = true,
+                     auth_domain = AuthDomain
+                    }};
+handle_auth_res(_, _, StateData) ->
+    send_element(StateData,
+                 #xmlel{name = <<"failure">>,
+                        attrs = [{<<"xmlns">>, ?NS_SASL}]}),
+    send_text(StateData, ?STREAM_TRAILER),
+    {stop, normal, StateData}.
+
+convert_decoded_cn(Val) when is_list(Val) ->
+    list_to_binary(Val);
+convert_decoded_cn(Val) when is_binary(Val) ->
+    Val;
+convert_decoded_cn(_) ->
+    error.
+
+get_ld_from_decoded_cn(error) ->
+    [];
+get_ld_from_decoded_cn(D) ->
+    case jid:from_binary(D) of
+        #jid{luser = <<"">>,
+             lserver = LD,
+             lresource = <<"">>} ->
+            [LD];
+        _ ->
+            []
+    end.
+
+get_ld_or_pcld_from_decoded_dsan({ok, SANs}) ->
```

Comment:
It is a cool name for a function, but I'd suggest making it a little bit more telling :)